### PR TITLE
Make delete or downgrade team owners scale

### PR DESCRIPTION
**Pull request by fisx**
https://github.com/zinfra/backend-issues/issues/1237

### CODE REVIEW

File: services/galley/src/Galley/API/Internal.hs [hunk 1]
```diff
@@ -63,7 +63,7 @@ rmUser user conn = do
   Data.eraseClients user
   where
     leaveTeams tids = for_ (result tids) $ \tid -> do
-      Data.teamMembersUnsafeForLargeTeams tid >>= uncheckedRemoveTeamMember user conn tid user
+      Data.teamMembersUnsafeForLargeTeams tid >>= uncheckedRemoveTeamMember user conn tid user . Just
```
Comment (mheinzel) [inline services/galley/src/Galley/API/Internal.hs:66]: we should not call `teamMembersUnsafeForLargeTeams` anymore, no?

Comment (fisx) [reply services/galley/src/Galley/API/Internal.hs:66]: Not part of this PR?

Comment (mheinzel) [reply services/galley/src/Galley/API/Internal.hs:66]: ah yes, it's not used to validate the deletion, but notify others about it.

File: services/brig/src/Brig/API/User.hs [hunk 2]
```diff
@@ -745,20 +744,10 @@ deleteUser uid pwd = do
     Nothing -> throwE DeleteUserInvalid
     Just a -> case accountStatus a of
       Deleted -> return Nothing
-      Suspended -> ensureNotOnlyOwner account >> go a
-      Active -> ensureNotOnlyOwner account >> go a
+      Suspended -> go a
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/brig/src/Brig/API/User.hs]: 8b4f360

File: services/galley/src/Galley/API/Error.hs [hunk 3]
```diff
@@ -112,13 +112,14 @@ actionDenied a =
...
+youMustBeOwnerWithEmail :: Error
+youMustBeOwnerWithEmail =
   Error
     status403
-    "no-other-owner"
+    "must-be-owner-with-email"
     "You are trying to remove or downgrade\
-    \ an owner. Promote another team member before proceeding."
+    \ an owner. You must own the team to do that, and you must\
+    \ have an email address associated with your user."
```
[the remark this answers is not in the document]
Comment (fisx) [reply services/galley/src/Galley/API/Error.hs]: should be resolve by 8b4f360.

**Comment by fisx** [comment]
what changed in error labels?
- when deleting `/self` (brig), "no-other-owner" changed to "no-self-delete-for-team-owner" (as its semantics also changed)
- when deleting or downgrading a team member (galley / team settings), "no-other-owner" changed to the already existing "access-denied", since it only is thrown when an admin deletes an owner.

**Comment by mheinzel** [comment]
I added more tests, some of which failed because we didn't guard against an owner deleting (`DELETE /teams/{tid}/members/{uid}`) or demoting itself. This is fixed now.

@fisx Please have a look at my changes tomorrow and merge if you're happy with them.

(I also had some search-related tests fail, but they seem unrelated, so I hope it's just a local thing and they'll pass in CI)

File: services/brig/src/Brig/API/Error.hs [hunk 4]
```diff
@@ -425,13 +425,12 @@ noBindingTeam = Wai.Error status403 "no-binding-team" "Operation allowed only on
 sameBindingTeamUsers :: Wai.Error
 sameBindingTeamUsers = Wai.Error status403 "same-binding-team-users" "Operation not allowed to binding team users."
 
-noOtherOwner :: Wai.Error
-noOtherOwner =
+deleteeIsOwner :: Wai.Error
+deleteeIsOwner =
```
Comment (akshaymankar) [inline services/brig/src/Brig/API/Error.hs:429]: This error is supposed to mean that the deletee and the deleter are both same person and owner, I think we should name it better, something like `ownerDeletingSelf`

File: services/galley/src/Galley/API/Teams.hs [hunk 5]
```diff
@@ -420,32 +422,45 @@ updateTeamMember zusr zcon tid targetMember = do
...
-  let privileged = filter (`canSeePermsOf` targetMember) updatedMembers
-      mkUpdate = EdMemberUpdate targetId
-      privilegedUpdate = mkUpdate $ Just targetPermissions
-      privilegedRecipients = membersToRecipients Nothing privileged
-  now <- liftIO getCurrentTime
-  let ePriv = newEvent MemberUpdate tid now & eventData ?~ privilegedUpdate
-  -- push to all members (user is privileged)
-  let pushPriv = newPush zusr (TeamEvent ePriv) $ privilegedRecipients
-  for_ pushPriv $ \p -> push1 $ p & pushConn .~ Just zcon
+  (updatedMembers, tooMany) <- Data.teamMembers' tid Nothing
```
Comment (akshaymankar) [inline services/galley/src/Galley/API/Teams.hs:437]: If we don't use `teamMembersUnsafeForLargeTeams` we risk sending wrong data to billing and the fanout. We have decided to not fanout things to larger teams, but can we please implement that when we're implementing that?

File: services/galley/src/Galley/API/Teams.hs [hunk 6]
```diff
@@ -464,13 +479,16 @@ deleteTeamMember zusr zcon tid remove mBody = do
...
+  zusrMember <- Data.teamMember tid zusr
+  targetMember <- Data.teamMember tid remove
+  void $ permissionCheck RemoveTeamMember zusrMember
+  do
+    dm <- maybe (throwM teamMemberNotFound) pure zusrMember
+    tm <- maybe (throwM teamMemberNotFound) pure targetMember
+    unless (canDeleteMember dm tm) $ throwM accessDenied
   team <- tdTeam <$> (Data.team tid >>= ifNothing teamNotFound)
   removeMembership <- Data.teamMember tid remove
-  mems <- Data.teamMembersUnsafeForLargeTeams tid
```
Comment (akshaymankar) [inline services/galley/src/Galley/API/Teams.hs:473]: Same  as above, can we deal with this separately please?

File: services/galley/src/Galley/API/Teams.hs [hunk 7]
```diff
@@ -427,22 +427,22 @@ updateTeamMember zusr zcon tid targetMember = do
       -- target user must be in same team
       throwM teamMemberNotFound
     Just previousMember -> do
-      okToDowngrade <- canDowngradeTeamMember user previousMember targetPermissions
-      unless okToDowngrade $ throwM accessDenied
+      when (downgradesOwner previousMember targetPermissions)
+        $
+        -- owner can be downgraded IFF it can be deleted
+        unless (canDeleteMember user previousMember)
+        $ throwM accessDenied
```
Comment (fisx) [inline services/galley/src/Galley/API/Teams.hs:434]: much nicer, thanks!

File: services/galley/src/Galley/Data.hs [hunk 8]
```diff
@@ -204,7 +205,10 @@ teamConversations t =
     <$> retry x1 (query Cql.selectTeamConvs (params Quorum (Identity t)))
 
 teamMembers :: forall m. (MonadThrow m, MonadClient m) => TeamId -> Range 1 HardTruncationLimit Int32 -> m ([TeamMember], Bool)
-teamMembers t (fromRange -> limit) = do
+teamMembers t = teamMembers' t . Just
+
+teamMembers' :: forall m. (MonadThrow m, MonadClient m) => TeamId -> Maybe (Range 1 HardTruncationLimit Int32) -> m ([TeamMember], Bool)
+teamMembers' t (maybe hardTruncationLimit fromRange -> limit) = do
```
Comment (akshaymankar) [inline services/galley/src/Galley/Data.hs:211]: All uses of this function send it `Nothing`, can we just call it `teamMembersWithDefaultLimit` and not take the `Maybe`?

I am a bit uncertain about adding a `'` to a function, it usually doesn't express intent very well. Unless I am missing some context.

File: services/galley/test/integration/API/SQS.hs [hunk 9]
```diff
@@ -100,7 +101,10 @@ tUpdate :: HasCallStack => Int32 -> [UserId] -> String -> Maybe E.TeamEvent -> I
 tUpdate c uids l (Just e) = do
   assertEqual (l <> "eventType") E.TeamEvent'TEAM_UPDATE (e ^. eventType)
   assertEqual "count" c (e ^. eventData . memberCount)
-  assertEqual "billing users" (toStrict . UUID.toByteString . toUUID <$> uids) (e ^. eventData . billingUser)
+  assertEqual
```
Comment (akshaymankar) [inline services/galley/test/integration/API/SQS.hs:104]: I implemented the same thing in the other PR! :smile:

File: services/galley/src/Galley/API/Teams.hs (hunk 7 above)
Comment (mheinzel) [reply services/galley/src/Galley/API/Teams.hs:434]: I don't like what ormolu did with the `$`, but I'll leave it now.

File: services/brig/src/Brig/API/Error.hs (hunk 4 above)
Comment (fisx) [reply services/brig/src/Brig/API/Error.hs:429]: I say it's actually about the owner role.  Let's discuss this.

**Pull request merged**